base/array.jl

```diff
@@ -2345,27 +2369,20 @@ julia> filter!(isodd, Vector(1:10))
 ```
 """
 function filter!(f, a::AbstractVector)
-    idx = eachindex(a)
-    y = iterate(idx)
-    y === nothing && return a
-    i, state = y
-
-    for acurr in a
-        if f(acurr)
-            @inbounds a[i] = acurr
-            y = iterate(idx, state)
-            y === nothing && (i += 1; break)
-            i, state = y
-        end
+    j = Int(firstindex(a))
+    for ai in a
+        @inbounds a[j] = ai
+        j = ifelse(f(ai), j+1, j)
```

Comment:
Is it valid to assume the indices are always consecutive integers?

Reply:
I would guess that it is legal, but you tell me. 

If it is not legal, then we'll need to use something like the second running iterator of the old code. This would be unfortunate, because it has the `y === nothing && (i += 1; break)` part that is unreachable (`eachindex(a)` cannot end before `for acurr in a`) and has an extra (always predicted) test+branch in the inner loop. We are at 2-4 cycles/item for cheap predicates, so this starts to matter.

Reply:
Wouldn't `nextind` work here?

Reply:
https://github.com/JuliaLang/julia/pull/21417

Reply:
If we can assume that `nextind` always works, then this perfectly solves the issue. 

When using the iteration protocol instead, I had some trouble getting the code fast and branchfree.